parsers/UY.py

```diff
@@ -1,41 +1,57 @@
 #!/usr/bin/python3
-
 import re
-from datetime import datetime
+from datetime import datetime, timedelta
 from logging import Logger, getLogger
-from typing import Optional
+from typing import Callable, List, Optional, Union
 
 import arrow
 import dateutil
-
-# BeautifulSoup is used to parse HTML to get information
 from bs4 import BeautifulSoup
 from requests import Session
 
+from parsers.lib.exceptions import ParserException
+
 tz = "America/Montevideo"
 
+# maps the xml keys from the website to names used by parser system
 MAP_GENERATION = {
-    "Hidráulica": "hydro",
-    "Eólica": "wind",
-    "Fotovoltaica": "solar",
-    "Biomasa": "biomass",
-    "Térmica": "oil",
+    # short for rio negro, main hydro power plant
+    "hydro": "r",
+    "wind": "eolica",
+    "solar": "fotovoltaica",
+    "biomass": "biomasa",
+    # all thermal, i.e. miscilaneous oil + natural gas production
+    "unknown": "termica",
+    "trade": "intercambios",
+    "demand": "demanda",
+    "salto_grande_prod": "comprassgu",
 }
-INV_MAP_GENERATION = dict([(v, k) for (k, v) in MAP_GENERATION.items()])
+
+AVALIABLE_KEYS = ["hydro", "wind", "solar", "biomass", "unknown"]
+
+UTE_URL = url = "https://ute.com.uy/energia-generada-intercambios-demanda"
 
 SALTO_GRANDE_URL = "http://www.cammesa.com/uflujpot.nsf/FlujoW?OpenAgent&Tensiones y Flujos de Potencia&"
 
 
-def get_salto_grande(session: Optional[Session]) -> float:
+def get_salto_grande(session: Session, targ_time: Optional[datetime] = None) -> float:
     """Finds the current generation from the Salto Grande Dam that is allocated to Uruguay."""
+    # handle optional arguments
+    s = session or Session()
+    lookup_time = datetime.now(tz=tz) if targ_time is None else targ_time
 
+    # Data for current hour seems to be available after 30mins.
+    # if we're too soon into the hour, check the hour before
     current_time = arrow.now("UTC-3")
-    if current_time.minute < 30:
-        # Data for current hour seems to be available after 30mins.
+    if (
+        current_time.floor("hour") == targ_time.floor("hour")
```

Comment:
Since `targ_time` can be undefined in it's current form calling `.floor()` on it generates a type error.

Reply:
This should be fixed in cb4aec7502c9d09a03148e912e5e29fbe2860044